README

                            raw2vmdk v0.1.3.2
                      by Tasos "Zapotek" Laskos
                          <tasos.laskos@gmail.com>

               Licensed under the GNU General Public License v2



General
----------
    raw2vmdk is an OS independent Java utility that allows you to mount
raw disk images, like images created by "dd", using VMware, VirtualBox or any
other virtualization platform supporting the VMDK disk format.

It analyzes the raw image and creates an approprietly formatted ".vmdk"
file that can be used to mount the image right away.

The ".vmdk" file is created based on the "vmdk.tpl" template file.

CAUTION:
    *Do not* remove or edit the vmdk.tpl file unless you are really
sure about what you're doing!


Features
-----------
A feature list follows:
    o Simple command line interface
    o No need to convert the image, run raw2vmdk and mount
        the created .vmdk file right away
    o OS independent


Usage
------

java -Dtype=<ide|buslogic|lsilogic|legacyESX> -jar raw2vmdk.jar <raw image> <vmdk outfile>

If no disk type has been specified it defaults to "ide".
The -Dtype parameter must be passed to the java VM as shown above,
not to raw2vmdk.

Note to MS Windows users:
    When passing the command line arguments escape all slashes in path names.

For example:
    java -jar raw2vmdk.jar C:\\images\\raw\\disk.img D:\\images\\vmdk\\disk.vmdk



Example
---------

zapotek@zaptop:~/raw2vmdk-0.1.jar$ java -jar raw2vmdk.jar "/media/disk-2/VM/FreeBSD 8.0 Minimal/freebsd8min.img" /home/zapotek/freebsd8min.vmdk
raw2vmdk 0.1.3.1 [$Rev: 28 $] initiated.
   Author: Zapotek <zapotek@segfault.gr>
   Website: http://www.segfault.gr

Analysing image:
/media/disk-2/VM/FreeBSD 8.0 Minimal/freebsd8min.img [1073741824 bytes]

Number of sectors:      2097152
Number of cylinders:    1023
Heads per track:        16
Sectors per track:      63

Loading VMDK template...
Writing VMDK file to: /home/zapotek/freebsd8min.vmdk
All done.


Requirements
-------------

* java version "1.6.0_18" or greater


Bug reports/Feature requests
-------------------------------
Please send your feedback using GitHub's issue system:
https://github.com/Zapotek/raw2vmdk/issues


License
---------
raw2vmdk is licensed under the GNU General Public License v2.
See the file "LICENSE" in the same folder for more informaton.


Disclaimer
-----------
raw2vmdk is free software and you are allowed to use it as you see fit.
However, I can't be held responsible for your actions or for any damage
caused by the use of this software.


CHANGELOG
Version 0.1.3.1
    * segfault/raw2vmdk/Raw2VMDK.java (main):
        fixed misleading example usage of the -Dtype parameter

Version 0.1.3
    * segfault/raw2vmdk/Raw2VMDK.java (main):
        added support for all available VMDK disk types

Version 0.1.2
	* segfault/raw2vmdk/Raw2VMDK.java (main):
	    replaced "\n" characters with System.getProperty( "line.separator" )

Version 0.1.1
    * segfault/raw2vmdk/VMDKTemplate.java (VMDTemplate):
        now throws exception if template file is not found

    * segfault/raw2vmdk/Raw2VMDK.java (main):
        handles exception thrown by VMDKTemplate.java

Version 0.1
    * First version.


src/segfault/raw2vmdk/VMDKTemplate.java

package segfault.raw2vmdk;

/*
 * VMDKTemplate.java Copyright (C) 2010 Tasos "Zapotek" Laskos <tasos.laskos@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 330, Boston, MA 02111-1307 USA
 */

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.HashMap;
import java.util.Iterator;

/**
 * <p>Template manager.</p>
 * <p>Loads the default VMDK template file, evaluates it and
 * creates new VMDK file based on {@link #write(HashMap, String)} parameters.</p>
 *
 * @author Tasos "Zapotek" Laskos <tasos.laskos@gmail.com>
 * @version 0.1.3.1
 */
public class VMDKTemplate {

    /**
     *  The VMDK template file
     */
    File tplFile;

    /**
     * Constructor <br/>
     * Initialises {@link #tplFile} using the tplLocation parameter
     *
     * @param tplLocation the location of the template file
     */
    public VMDKTemplate( String tplLocation ) throws Exception {

        this.tplFile = new File( tplLocation );

        if( !tplFile.exists( ) ) {
            throw new Exception( "Template file does not exist." );
        }
    }

    /**
     * Creates new VMDK file "outFile" from the template in "tpl" with the values
     * of tplData
     *
     * @param tplData   the template data data in a tplVarName => value
     * @param outFile   where to write the .vmdk file
     */
    public void write( HashMap<String, String> tplData, String outFile ) {

        try {
//            File tplFile = new File( tpl );
            BufferedReader reader = new BufferedReader(
                    new FileReader( tplFile ) );

            String line = "", tplText = "";

            // read the template file and store it in vmdkTpl
            while( ( line = reader.readLine( ) ) != null ) {
                tplText += line + System.getProperty( "line.separator" );
            }

            // System.out.println( tplText );
            reader.close( );

            Iterator<String> it = tplData.keySet( ).iterator( );
            String vmdkText = tplText;

            // replace template variables with actual values
            while( it.hasNext( ) ) {
                String key = (String) it.next( );
                String val = tplData.get( key );

                // System.out.println( key + "::" + val );

                vmdkText = vmdkText.replaceAll( "\\[" + key + "\\]", val );

            }

            // System.out.println( "vmdkText: " + vmdkText );

            FileWriter writer = new FileWriter( outFile );
            writer.write( vmdkText );
            writer.close( );

        } catch( IOException ioe ) {
            ioe.printStackTrace( );
        }
    }

}


src/cert/forensics/mbr/PartitionEntry.java
/*
 * PartitionEntry.java Copyright (C) 2006-2008 Carnegie Mellon University
 * 
 * Tim Vidas <tvidas at gmail d0t com> Brian Kaplan <bfkaplan at cmu d0t edu>
 * 
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 330, Boston, MA 02111-1307 USA
 */

package cert.forensics.mbr;

/**
 * ParitionEntry Represents one of the four partition entries in an MBR
 * 
 * @author Tim Vidas
 * @author Brian Kaplan
 * @version 0.7, Jan 2009
 */

public class PartitionEntry {

    private static final int PARTITION_ENTRY_SIZE = 16;

    private int[]            entryBytes;

    // known type signatures to sanity check validity of the partition entry and
    // mbr
    private int[]            validPartitionTypes  = { 0x00, 0x10, 0x02, 0x03,
            0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0e, 0x0f,
            0x10, 0x11, 0x12, 0x14, 0x16, 0x17, 0x18, 0x1b, 0x1c, 0x1e, 0x24,
            0x39, 0x3c, 0x40, 0x41, 0x42, 0x4d, 0x4e, 0x4f, 0x50, 0x51, 0x52,
            0x53, 0x54, 0x55, 0x56, 0x5c, 0x61, 0x63, 0x64, 0x65, 0x70, 0x75,
            0x80, 0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88, 0x8e, 0x93,
            0x94, 0x9f, 0xa0, 0xa5, 0xa6, 0xa7, 0xa8, 0xa9, 0xab, 0xb7, 0xb8,
            0xbb, 0xbe, 0xbf, 0xc1, 0xc4, 0xc6, 0xc7, 0xda, 0xdb, 0xde, 0xdf,
            0xe1, 0xe3, 0xe4, 0xeb, 0xee, 0xef, 0xf0, 0xf1, 0xf4, 0xf2, 0xfd,
            0xfe, 0xff                           };

    // structure of parititon table
    private int              state;                    // byte 1

    private int              beginHead;                // byte 2

    private int              beginCylinderSector;      // byte 3+4

    private int              partitionType;            // byte 5

    private int              endHead;                  // byte 6

    private int              numHeads;

    private int              numCylinders;

    private int              endCylinderSector;        // byte 7+8

    private long             relativeSector;           // byte 9+10+11+12

    // (sectors between mbr
    // and 1st sector of
    // parition)

    private long             numSectors;               // byte 13+14+15+16

    /**
     * constructor
     * 
     * @param pe
     *            integer array of the partition entry
     */
    public PartitionEntry( int[] pe ) {

        if( pe.length == PARTITION_ENTRY_SIZE )
            entryBytes = pe;
        else
            entryBytes = new int[ PARTITION_ENTRY_SIZE ];

        state = entryBytes[0];
        beginHead = entryBytes[1];
        beginCylinderSector = ( ( entryBytes[3] << 8 ) | entryBytes[2] ); // combine
        // bytes
        // 3 +
        // 2
        // (backwards)
        partitionType = entryBytes[4];
        endHead = entryBytes[5];
        endCylinderSector = ( ( entryBytes[7] << 8 ) | entryBytes[6] ); // combine
        // bytes
        // 7 + 6
        // (backwards)

        relativeSector = ( ( entryBytes[11] & 0xFF ) << 24 ) // combine bytes
                // 8-11 (backwards)
                | ( ( entryBytes[10] & 0xFF ) << 16 )
                | ( ( entryBytes[9] & 0xFF ) << 8 ) | ( entryBytes[8] & 0xFF );

        numSectors = ( ( entryBytes[15] & 0xFF ) << 24 ) // combine bytes 12-15
                // (backwards)
                | ( ( entryBytes[14] & 0xFF ) << 16 )
                | ( ( entryBytes[13] & 0xFF ) << 8 ) | ( entryBytes[12] & 0xFF );

        setNumHeads( beginHead + endHead );
        numCylinders = getCylinder( beginCylinderSector )
                + getCylinder( endCylinderSector );

    }

    /**
     * Extract 6 bit sector from Cylinder/Sector 16bit structure
     * 
     * @param cylSectStructure
     *            the cylinder Sector Structure
     * @return just the sector part of cylSectStructure
     */
    private static int getSector( int cylSectStructure ) {

        return( cylSectStructure & 63 );
    }

    /**
     * Extract 10bit cylinder from Cylinder/Sector 16bit structure
     * 
     * @param cylSectStructure
     *            the cylinder Sector Structure
     * @return just the cylinder part of cylSectStructure
     */
    private static int getCylinder( int cylSectStructure ) {

        int bits8To15 = ( ( cylSectStructure & 65280 ) >> 8 );

        int cylinderVal = bits8To15;

        if( ( cylSectStructure & 64 ) != 0 ) // if the 6th bit of the 16 bit
            // structure is not zero
            cylinderVal += 256; // set bit 8 of the 10 bit cylinder value

        if( ( cylSectStructure & 128 ) != 0 ) // if the 7th bit of the 16 bit
            // structure is not zero
            cylinderVal += 512; // set bit 9 of the 10 bit cylinder value

        return cylinderVal;
    }

    /**
     * gets the beginning cylinder
     * 
     * @return the beginning cylinder
     */
    public int getBeginCylinder( ) {

        return getCylinder( beginCylinderSector );
    }

    /**
     * gets the beginning sector
     * 
     * @return the beginning sector
     */
    public int getBeginSector( ) {

        return getSector( beginCylinderSector );
    }

    /**
     * inspector for beginHead
     * 
     * @return the beginHead datamember
     */
    public int getBeginHead( ) {

        return beginHead;
    }

    /**
     * gets the end cylinder
     * 
     * @return the end cylinder
     */
    public long getEndCylinder( ) {

        return getCylinder( endCylinderSector );
    }

    /**
     * gets the end sector
     * 
     * @return the end sector
     */
    public long getEndSector( ) {

        return getSector( endCylinderSector );
    }

    /**
     * inspector for endHead
     * 
     * @return the endHead datamember
     */
    public int getEndHead( ) {

        return endHead;
    }

    /**
     * inspector for entryBytes
     * 
     * @return the entryBytes datamember
     */
    public int[] getEntryBytes( ) {

        return entryBytes;
    }

    /**
     * inspector for numSectors
     * 
     * @return the numSectors datamember
     */
    public long getNumSectors( ) {

        return numSectors;
    }

    /**
     * inspector for partitionType
     * 
     * @return the partitionType datamember
     */
    public int getPartitionType( ) {

        return partitionType;
    }

    /**
     * inspector for relativeSector
     * 
     * @return the relativeSector datamember
     */
    public long getRelativeSector( ) {

        return relativeSector;
    }

    /**
     * inspector for state
     * 
     * @return the state datamember
     */
    public int getState( ) {

        return state;
    }

    /**
     * checks to see if this partition is bootable (has 0x80 state)
     * 
     * @return true if bootable, fales otherwise
     */
    public boolean isBootable( ) {

        return state == 0x80; // flag 0x80 means partition is bootable
    }

    /**
     * checks if the partition type is a "known" value so it is really most
     * useful for checking if something is not valid
     * 
     * @return true if partition is valid, false otherwise
     */
    public boolean isValidPartition( ) {

        for( int i = 0; i < validPartitionTypes.length; i++ ) {
            if( validPartitionTypes[i] == partitionType )
                return true;
        }
        return false;
    }

    /**
     * Really only used to check bootable partitions, so not all obscure windows
     * partition type flags are checked
     * 
     * @return true if it is a known windows partition type, false otherwise
     */
    public boolean isNotWindowsBased( ) {

        if( partitionType == 0x07 || // NTFS
                partitionType == 0x0b || // FAT32 CHS
                partitionType == 0x0c || // FAT32 LBA
                partitionType == 0x06 // FAT16

        )
            return false; // it may be windows based
        else
            return true; // can't be windows based
    }

    /**
     * checks if this partition is FAT
     * 
     * @return true is it's FAT, false otherwise
     */
    public boolean isFAT( ) {

        return( partitionType == 0x0b || partitionType == 0x0c || partitionType == 0x06 );
    }

    /**
     * checks if this partition is NTFS
     * 
     * @return true is it's NTFS, false otherwise
     */
    public boolean isNTFS( ) {

        return( partitionType == 0x07 );
    }

    /**
     * generic toString method that assembles datamembers
     * 
     * @return a formated string
     */
    public String toString( ) {

        StringBuffer sb = new StringBuffer( );
        sb.append( "Is Bootable: " + isBootable( )
                + System.getProperty( "line.separator" ) );
        sb.append( "Begin Head: " + beginHead
                + System.getProperty( "line.separator" ) );
        sb.append( "Begin Cylinder: " + getCylinder( beginCylinderSector )
                + System.getProperty( "line.separator" ) );
        sb.append( "Begin Sector: " + getSector( beginCylinderSector )
                + System.getProperty( "line.separator" ) );
        sb.append( "Partition Type: " + "0x"
                + Integer.toHexString( partitionType )
                + System.getProperty( "line.separator" ) );
        sb.append( "End Head: " + endHead
                + System.getProperty( "line.separator" ) );
        sb.append( "End Cylinder: " + getCylinder( endCylinderSector )
                + System.getProperty( "line.separator" ) );
        sb.append( "End Sector: " + getSector( endCylinderSector )
                + System.getProperty( "line.separator" ) );
        sb.append( "Relative Sector: " + relativeSector
                + System.getProperty( "line.separator" ) );
        sb.append( "Num Sectors: " + numSectors
                + System.getProperty( "line.separator" ) );
        sb.append( "Num Heads: " + getNumHeads( )
                + System.getProperty( "line.separator" ) );
        sb.append( "Num Cylinders: " + numCylinders
                + System.getProperty( "line.separator" ) );

        return sb.toString( );
    }

    /**
     * @param numHeads
     *            the numHeads to set
     */
    public void setNumHeads( int numHeads ) {

        this.numHeads = numHeads;
    }

    /**
     * @return the numHeads
     */
    public int getNumHeads( ) {

        return numHeads;
    }

}


src/cert/forensics/mbr/MasterBootRecord.java

package cert.forensics.mbr;

/*
 * MasterBootRecord.java Copyright (C) 2006-2008 Carnegie Mellon University
 * 
 * Tim Vidas <tvidas at gmail d0t com> Brian Kaplan <bfkaplan at cmu d0t edu>
 * 
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 330, Boston, MA 02111-1307 USA
 */

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;

/**
 * MasterBootRecord Represents a 512 byte Master Boot Record for a disk
 * 
 * @author Tim Vidas
 * @author Brian Kaplan
 * @version 0.7, Jan 2009
 */

public class MasterBootRecord {

    private static final int BOOT_CODE_SIZE   = 446;

    private static final int MARKER_SIZE      = 2;

    private static final int BYTES_PER_SECTOR = 512;

    private int[]            bootCode;

    public PartitionEntry    partitionEntry1;

    private PartitionEntry   partitionEntry2;

    private PartitionEntry   partitionEntry3;

    private PartitionEntry   partitionEntry4;

    private int[]            marker;

    private long             fileSizeBytes;

    private int[]            mbrUnsignedBytes;

    /**
     * constructor for MasterBootRecord
     * 
     * @param image
     *            File to a disk image to obtain the mbr from
     */
    public MasterBootRecord( File image ) {

        mbrUnsignedBytes = new int[ 512 ];

        setFileSizeBytes( image.length( ) );

        // read contents of mbr into a 512b mbr buffer
        try {
            InputStream imageStream = new FileInputStream( image );

            /** read mbr into buffer **/
            for( int i = 0; i < mbrUnsignedBytes.length; i++ )
                mbrUnsignedBytes[i] = imageStream.read( );
        } catch( IOException ioe ) {
            System.out.println( "Problem: " + ioe );
        }

        initialize( mbrUnsignedBytes );
    }

    /**
     * constructor for MasterBootRecord
     * 
     * @param mbrUnsignedBytes
     *            a boot record as bytes
     */
    public MasterBootRecord( int[] mbrUnsignedBytes ) {

        initialize( mbrUnsignedBytes );
    }

    /**
     * initializes datamembers from bytes representing an MBR
     * 
     * @param mbrUnsignedBytes
     *            a boot record as bytes
     */
    private void initialize( int[] mbrUnsignedBytes ) {

        this.mbrUnsignedBytes = mbrUnsignedBytes;
        // carve out CODE_SIZE bytes from image
        bootCode = new int[ BOOT_CODE_SIZE ];
        for( int i = 0; i < BOOT_CODE_SIZE; i++ )
            bootCode[i] = mbrUnsignedBytes[i];

        // carve out Partition Entries from image
        int[] pe1 = new int[ 16 ];
        int[] pe2 = new int[ 16 ];
        int[] pe3 = new int[ 16 ];
        int[] pe4 = new int[ 16 ];

        for( int i = BOOT_CODE_SIZE; i < BOOT_CODE_SIZE + 16; i++ ) // loop
        // through 4
        // partition
        // tables
        // (16 bytes
        // each)
        {
            pe1[i - BOOT_CODE_SIZE] = mbrUnsignedBytes[i]; // copy byte from mbr
            // to partition entry
            // buffer 1
            pe2[i - BOOT_CODE_SIZE] = mbrUnsignedBytes[i + 16]; // copy byte
            // from mbr to
            // partition
            // entry buffer
            // 2
            pe3[i - BOOT_CODE_SIZE] = mbrUnsignedBytes[i + 32]; // copy byte
            // from mbr to
            // partition
            // entry buffer
            // 3
            pe4[i - BOOT_CODE_SIZE] = mbrUnsignedBytes[i + 48]; // copy byte
            // from mbr to
            // partition
            // entry buffer
            // 4
        }
        partitionEntry1 = new PartitionEntry( pe1 );
        partitionEntry2 = new PartitionEntry( pe2 );
        partitionEntry3 = new PartitionEntry( pe3 );
        partitionEntry4 = new PartitionEntry( pe4 );

        // carve out marker from image
        marker = new int[ MARKER_SIZE ];

        for( int i = BOOT_CODE_SIZE + 64; i < BOOT_CODE_SIZE + 64 + 2; i++ ) {
            marker[i - BOOT_CODE_SIZE - 64] = mbrUnsignedBytes[i];
        }
    }

    /**
     * inspector for bootCode
     * 
     * @return the bootCode datamember
     */
    public int[] getBootCode( ) {

        return bootCode;
    }

    /**
     * inspector for marker
     * 
     * @return the marker datamember
     */
    public int[] getMarker( ) {

        return marker;
    }

    /**
     * determines a bootable partition
     * 
     * @return the bootable PartitionEntry
     */
    public PartitionEntry getBootablePartition( ) {

        if( partitionEntry1.isBootable( ) )
            return partitionEntry1;
        else if( partitionEntry2.isBootable( ) )
            return partitionEntry2;
        else if( partitionEntry3.isBootable( ) )
            return partitionEntry3;
        else if( partitionEntry4.isBootable( ) )
            return partitionEntry4;
        else
            return null;
    }

    /**
     * determines bootable partition index
     * 
     * @return the id of the bootable partition (1-4)
     */
    public int getBootablePartitionIndex( ) {

        if( partitionEntry1.isBootable( ) )
            return 1;
        else if( partitionEntry2.isBootable( ) )
            return 2;
        else if( partitionEntry3.isBootable( ) )
            return 3;
        else if( partitionEntry4.isBootable( ) )
            return 4;
        else
            return -1;
    }

    /**
     * inspector for ParitionEntry1
     * 
     * @return the PartitionEntry1 datamember
     */
    public PartitionEntry getPartitionEntry1( ) {

        return partitionEntry1;
    }

    /**
     * inspector for ParitionEntry2
     * 
     * @return the PartitionEntry2 datamember
     */
    public PartitionEntry getPartitionEntry2( ) {

        return partitionEntry2;
    }

    /**
     * inspector for ParitionEntry3
     * 
     * @return the PartitionEntry3 datamember
     */
    public PartitionEntry getPartitionEntry3( ) {

        return partitionEntry3;
    }

    /**
     * inspector for ParitionEntry4
     * 
     * @return the PartitionEntry4 datamember
     */
    public PartitionEntry getPartitionEntry4( ) {

        return partitionEntry4;
    }

    /**
     * get total sectors reported by all partitions on disk
     * 
     * Note on some drives the Vista installer may actually set this value too
     * large It has been observed on 13.5 GB drives
     * 
     * @return the total number of sectors
     */
    public long totalSectorsFromPartitions( ) {

        return partitionEntry1.getNumSectors( )
                + partitionEntry2.getNumSectors( )
                + partitionEntry3.getNumSectors( )
                + partitionEntry4.getNumSectors( );
    }

    /**
     * use disk image file size to calculate total sectors
     * 
     * @return total number of sectors
     */
    public long totalSectorsOnDiskFromFile( ) {

        return getFileSizeBytes( ) / BYTES_PER_SECTOR;
    }

    /**
     * used for getting the "end cylinder" value for a disk
     * 
     * @return the end cylinder
     */
    public long largestCylinderValOnDisk( ) {

        // return the maximum end cylinder value for the four partitions
        return Math.max( Math.max( partitionEntry1.getEndCylinder( ),
                partitionEntry2.getEndCylinder( ) ), Math.max( partitionEntry3
                .getEndCylinder( ), partitionEntry4.getEndCylinder( ) ) );
    }

    /**
     * Checks if the MBR is valid by checking if there is a bootable partition
     * and if the bootable partition is valid (of a known type)
     * 
     * @return true if there is a bootable partition, false otherwise
     */
    public boolean isValidMBR( ) {

        if( getBootablePartition( ) == null ) // if there is no bootable
            // partition (0x80 flag)
            return false;

        if( !getBootablePartition( ).isValidPartition( ) ) // if bootable is not
            // a valid partition
            return false;

        return true;
    }

    /**
     * Checks for the 33 C0 8E starting bootcode common to known Windows Systems
     * 
     * @return true - appears to be a windows bootcode, false otherwise
     */
    public boolean hasWindowsBootcode( ) {

        return( bootCode[0] == 0x33 && bootCode[1] == 0xC0 && bootCode[2] == 0x8E );
    }

    /**
     * generic toString method that assembles datamembers
     * 
     * @return a formated string
     */
    public String toString( ) {

        StringBuffer sb = new StringBuffer( );
        for( int i = 0; i < mbrUnsignedBytes.length; i++ ) {
            if( i % 16 == 0 )
                sb.append( System.getProperty( "line.separator" ) );
            if( mbrUnsignedBytes[i] < 16 )
                sb.append( "0" + Integer.toHexString( mbrUnsignedBytes[i] )
                        + " " ); // pad 0-F with leading 0
            else
                sb.append( Integer.toHexString( mbrUnsignedBytes[i] ) + " " ); // print
            // two
            // char
            // hex
            // val
            // 0-255
        }
        sb.append( System.getProperty( "line.separator" ) );
        sb.append( "Partition 1:" + System.getProperty( "line.separator" ) );
        sb.append( "===================="
                + System.getProperty( "line.separator" ) );
        sb.append( partitionEntry1.toString( ) );
        sb.append( "Partition 2:" + System.getProperty( "line.separator" ) );
        sb.append( "===================="
                + System.getProperty( "line.separator" ) );
        sb.append( partitionEntry2.toString( ) );
        sb.append( "Partition 3:" + System.getProperty( "line.separator" ) );
        sb.append( "===================="
                + System.getProperty( "line.separator" ) );
        sb.append( partitionEntry3.toString( ) );
        sb.append( "Partition 4:" + System.getProperty( "line.separator" ) );
        sb.append( "===================="
                + System.getProperty( "line.separator" ) );
        sb.append( partitionEntry4.toString( ) );
        sb.append( System.getProperty( "line.separator" )
                + System.getProperty( "line.separator" ) );

        sb.append( "Total sectors: " + totalSectorsOnDiskFromFile( )
                + System.getProperty( "line.separator" ) );
        sb.append( "Total cylinders: " + largestCylinderValOnDisk( ) );

        return sb.toString( );
    }

    /**
     * @param fileSizeBytes
     *            the fileSizeBytes to set
     */
    public void setFileSizeBytes( long fileSizeBytes ) {

        this.fileSizeBytes = fileSizeBytes;
    }

    /**
     * @return the fileSizeBytes
     */
    public long getFileSizeBytes( ) {

        return fileSizeBytes;
    }
}


src/segfault/raw2vmdk/Raw2VMDK.java

package segfault.raw2vmdk;

/*
 * Raw2VMDK.java Copyright (C) 2010 Tasos "Zapotek" Laskos <tasos.laskos@gmail.com>
 *
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 330, Boston, MA 02111-1307 USA
 */

import java.io.File;
import java.util.*;

import cert.forensics.mbr.MasterBootRecord;

/**
 * <p>Main class.</p>
 * <p>Drives the cert.forensics.mbr.MasterBootRecord and
 * segfault.raw2vmdk.VMDKTemplate classes.</p>
 *
 * @author Tasos "Zapotek" Laskos <tasos.laskos@gmail.com>
 * @version 0.1.3.2
 *
 * @see cert.forensics.mbr.MasterBootRecord
 * @see segfault.raw2vmdk.VMDKTemplate
 */
public class Raw2VMDK {

    /**
     * The version of the application.<br/>
     * Used in {@link #banner()}
     */
    private static final String VERSION = "0.1.3.2";

    /**
     * The SVN revision of the application.<br/>
     * Used in {@link #banner()}
     */
    private static final String SVN_REV = "$Rev$";

    /**
     * total number of sectors
     */
    static long                 numOfSectors;

    /**
     * total number of cylinders
     */
    static long                 numOfCylinders;

    /**
     *  heads per track
     */
    static int                  headsPerTrack;

    /**
     *  sectors per track
     */
    static long                 sectorsPerTrack;

    /**
     * the VMDK template
     */
    static String               tpl     = "vmdk.tpl";

    /**
     *  the location of the raw image
     */
    static String               imageLocation;

    /**
     * <p>Main method</p>
     * <p>Drives the cert.forensics.mbr.MasterBootRecord and
     * segfault.raw2vmdk.VMDKTemplate classes.</p>
     *
     * @param args  <raw image> <vmdk outfile>
     *
     * @see cert.forensics.mbr.MasterBootRecord
     * @see segfault.raw2vmdk.VMDKTemplate
     */
    public static void main( String[] args ) {

        banner( );

        if( args.length == 0 ) {
            usage( );
            return;
        }

        if( args.length < 2 ) {
            usage( );
            System.out.println( System.getProperty( "line.separator" )
                    + "Error: raw2vmdk expects at least 2 arguments." );
            return;
        }

        String diskType = System.getProperty( "type", "ide" );
        String[] acceptedTypes = { "ide", "buslogic", "lsilogic", "legacyESX" };

        if( !inArray( acceptedTypes, diskType ) ) {
            System.out.println( System.getProperty( "line.separator" )
                    + "Error: Disk type is incorrect." );
            usage( );
            return;
        }

        imageLocation   = args[0];

        File imgFile    = new File( imageLocation );

        // check if the raw image file exists
        if( !imgFile.exists( ) ) {
            System.out.println( System.getProperty( "line.separator" )
                    + "Error: Image file does not exist." );
            return;
        }

        System.out.print( "Analysing image:"
                + System.getProperty( "line.separator" ) + imageLocation );

        // analyse the image
        MasterBootRecord MBR = new MasterBootRecord( imgFile );

        System.out.println( " [" + MBR.getFileSizeBytes( ) + " bytes]" );
        System.out.println( );

        numOfSectors = MBR.totalSectorsOnDiskFromFile( );
        System.out.println( "Number of sectors:\t" + numOfSectors );

        numOfCylinders = MBR.largestCylinderValOnDisk( );
        System.out.println( "Number of cylinders:\t" + numOfCylinders );

        headsPerTrack = MBR.getPartitionEntry1( ).getNumHeads( );
        System.out.println( "Heads per track:\t" + headsPerTrack );

        sectorsPerTrack = MBR.getPartitionEntry1( ).getEndSector( );
        System.out.println( "Sectors per track:\t" + sectorsPerTrack );

        // create hashmap holding data for the VMDK template
        HashMap<String, String> vmdkData = new HashMap<String, String>( );

        vmdkData.put( "diskType", diskType );
        vmdkData.put( "numOfSectors", Long.toString( numOfSectors ) );
        vmdkData.put( "numOfCylinders", Long.toString( numOfCylinders ) );
        vmdkData.put( "headsPerTrack", Integer.toString( headsPerTrack ) );
        vmdkData.put( "sectorsPerTrack", Long.toString( sectorsPerTrack ) );
        vmdkData.put( "imgLocation", imageLocation );

        System.out.println( System.getProperty( "line.separator" )
                + "Loading VMDK template..." );

        try {
            // load VMDK template file
            VMDKTemplate vmdkTpl = new VMDKTemplate( tpl );

            System.out.print( "Writing VMDK file to: " );

            // write VMDK file to disk
            vmdkTpl.write( vmdkData, args[1] );
            System.out.println( args[1] );
        } catch( Exception e ) {
            System.out.println( System.getProperty( "line.separator" )
                    + "Error: " + e.getMessage( ) );
            return;
        }

        System.out
                .println( "All done." + System.getProperty( "line.separator" ) );
    }

    /**
     * Outputs the banner message of the application
     */
    public static void banner( ) {

        System.out.println( "raw2vmdk " + VERSION + " [" + SVN_REV
                + "] initiated." );
        System.out.println( "   Author: Tasos \"Zapotek\" Laskos <tasos.laskos@gmail.com>" );
        System.out.println( "   Website: http://www.segfault.gr" );
        System.out.println( );
    }

    /**
     * Outputs usage info
     */
    public static void usage( ) {

        System.out.println( );
        System.out.println( "Usage:" );
        System.out
            .println( "java -jar -Dtype=<ide|buslogic|lsilogic|legacyESX> raw2vmdk.jar <raw image> <vmdk outfile>" );

        System.out.println( "\ntype defaults to 'ide'" );
    }

    private static Boolean inArray( String[] arr, String str ) {

        for( String s: arr ) {
          if( s.equals( str ) ) return true;
        }

        return false;

    }

}
